File: 37-sudoku-solver/sudoku-solver.java
class Solution {
    char[][] board;
    
    boolean isValidPlace(int x,int y,char val){
        for(int i=0;i<9;i++){
            if(board[x][i]==val){
                return false;
            }
            if(board[i][y]==val){
                return false;
            }
        }
        

        int startX = (x/3) * 3;
        int startY = (y/3) * 3;
        for(int i=0;i<3;i++){
            for(int j=0;j<3;j++){
                if(board[startX+i][startY+j]==val){
                    return false;
                }
            }
        }
        return true;
    }
    boolean solve(){
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                if(board[i][j]=='.'){
                    for(char c='1';c<='9';c++){
                        if(isValidPlace(i,j,c)){
                            board[i][j]=c;
                            if(solve()) return true;
                            board[i][j]='.'; 
                        }
                    }
                    return false; 
                }
            }
        }
        return true; 
    }
    public void solveSudoku(char[][] board) {
        this.board=board;
        solve();
    }
}

File: 37-sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
      
        rows=[set() for _ in range(9)]
        cols=[set() for _ in range(9)]
        boxes=[set() for _ in range(9)]
        empty_cells=[]

        for r in range(9):
            for c in range(9):
                val=board[r][c]
                if board[r][c]=='.':
                    empty_cells.append((r,c))
                else:
                    rows[r].add(val)
                    cols[c].add(val)
                    box_index=(r//3)*3+(c//3)
                    boxes[box_index].add(val)
        
        def backtrack(index:int)->bool:
            if index==len(empty_cells):
                return True
            
            r,c=empty_cells[index]
            box_index=(r//3)*3+(c//3)

            for num in map(str,range(1,10)):
                if num not in rows[r] and num not in cols[c] and num not in boxes[box_index]:
                    board[r][c]=num
                    rows[r].add(num)
                    cols[c].add(num)
                    boxes[box_index].add(num)

                    if backtrack(index+1):
                        return True
                    
                    board[r][c]='.'
                    rows[r].remove(num)
                    cols[c].remove(num)
                    boxes[box_index].remove(num)
            return False
        backtrack(0)